File: LeetCode/0019.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution
{
public:
    ListNode *removeNthFromEnd(ListNode *head, int n)
    {
        ListNode *newhead = new ListNode(0);
        newhead->next = head;
        ListNode *fast = newhead, *low = newhead;

        while (n--)
            fast = fast->next;
        while (fast->next)
        {
            fast = fast->next;
            low = low->next;
        }

        low->next = low->next->next;
        return newhead->next;
    }
};

File: LeetCode/0019.go
package leetcode

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func removeNthFromEnd(head *ListNode, n int) *ListNode {
	dummynode := &ListNode{Val: 0}
	dummynode.Next = head
	fast, low := dummynode, dummynode
	for ; n > 0; n-- {
		fast = fast.Next
	}
	for fast.Next != nil {
		fast = fast.Next
		low = low.Next
	}
	low.Next = low.Next.Next
	return dummynode.Next
}
